LOOP-1114: calls /info API to do a version check

rickpasetto opened the pull request:
https://tidepool.atlassian.net/browse/LOOP-1114

rickpasetto committed:
ckpt

rickpasetto committed:
LOOP-1114: calls /info API to do a version check

https://tidepool.atlassian.net/browse/LOOP-1114

darinkrauss commented on TidepoolServiceKit/TidepoolService.swift:
```diff
@@ -233,6 +233,26 @@ extension TidepoolService: RemoteDataService {
 
+extension TidepoolService: VersionCheckService {
+    public func checkVersion(currentVersion: String, completion: @escaping (Result<VersionUpdate, Error>) -> Void) {
+        tapi.getInfo(environment: nil) { result in
```
[nit] Since the default is `nil`, just leave environment parameter off.

darinkrauss commented on TidepoolServiceKit/TidepoolService.swift:
```diff
@@ -233,6 +233,26 @@ extension TidepoolService: RemoteDataService {
 }
 
+extension TidepoolService: VersionCheckService {
+    public func checkVersion(currentVersion: String, completion: @escaping (Result<VersionUpdate, Error>) -> Void) {
```
See my comment in related LoopKit PR about the `bundleIdentifier` parameter and checking against hard-coded Tidepool Loop bundle identifier.

darinkrauss reviewed (changes requested):
See my comments and requested changes in LoopKit.

rickpasetto committed:
PR Feedback

rickpasetto committed:
PR Feedback

darinkrauss commented on TidepoolServiceKit/LoopVersionInfo.swift:
```diff
@@ -0,0 +1,52 @@
+        self.criticalUpdateNeeded = criticalUpdateNeeded
+    }
+    
+    public func getVersionUpdateNeeded(currentVersion version: String) -> VersionUpdate {
```
[nit] Why not just add this functionality as a local extension to `TInfo.Versions.Loop`?

rickpasetto replied:
I was viewing `TInfo` as policy-free, but `TidepoolService` as having the policy for this.

darinkrauss replied:
Right, but that doesn't mean you can't extend `TInfo` privately for the `TidepoolService`, right?

rickpasetto replied:
ok

darinkrauss commented on TidepoolServiceKitTests/SemanticVersionTests.swift:
```diff
@@ -0,0 +1,36 @@
+        XCTAssertNil(SemanticVersion("abc123"))
+        XCTAssertNil(SemanticVersion("foopyNoopy"))
+        XCTAssertNil(SemanticVersion("1.2.3.4"))
+        XCTAssertNil(SemanticVersion("-1.2.3.4"))
```
[nit] Is this going to assert while trying to run the unit tests?

rickpasetto replied:
Nope, because it won't match the regex

darinkrauss replied:
Ah, then why the `assert(major >= 0)`s?

rickpasetto replied:
In case someone changes the regex

darinkrauss reviewed (approved):
A couple of nits, but otherwise LGTM!

rickpasetto committed:
PR Feedback: adds persistence to the last version info

Also added unit tests for TidepoolService.checkVersion().
Note: copied URLProtocolMock from TidepoolKit tests

rickpasetto committed:
Fix name of VersionInfoTests

rickpasetto committed:
Gets rid of unnecessary code

rickpasetto committed:
More PR Feedback

darinkrauss reviewed (approved):
LGTM! Thanks for making those changes.
